Feat: Apertus model implementation

pwilkin opened the pull request:
Help!

This seems like an easy enough model. I've done the XIELU implementation, but for some reason, the tensor outputs diverge in the QKV calculations (it's either the RoPE or the norm). I've tried everything, spent a whole day debugging it, maybe I'm missing something really obvious?

@foldl you did the implementation in chatllm.cpp and it works fine, maybe you can see something obvious here?

pwilkin committed:
First attempt

pwilkin commented:
Would solve https://github.com/ggml-org/llama.cpp/issues/15748

foldl commented:
@pwilkin you can check rope-type and the soft-plus thing.

pwilkin commented:
Rope type is fine I thing (normal, not NeoX), softplus doesn't matter as the problem is before the xIELU activation (first tensor matches reference through all 30 layers, next ones diverge starting from first kqv-out).

pwilkin commented:
Norm application is wrong, but I totally can't figure out how exactly :/

ggerganov commented:
Which norm?

pwilkin commented:
@ggerganov KQ norm.

ggerganov commented:
I think forgot to apply the norm bias?

pwilkin commented:
@ggerganov That's the thing, it doesn't have a norm bias:

```json
  "weight_map": {
    "lm_head.weight": "model-00004-of-00004.safetensors",
    "model.embed_tokens.weight": "model-00001-of-00004.safetensors",
    "model.layers.0.attention_layernorm.weight": "model-00001-of-00004.safetensors",
    "model.layers.0.feedforward_layernorm.weight": "model-00001-of-00004.safetensors",
    "model.layers.0.mlp.act_fn.alpha_n": "model-00001-of-00004.safetensors",
    "model.layers.0.mlp.act_fn.alpha_p": "model-00001-of-00004.safetensors",
    "model.layers.0.mlp.act_fn.beta": "model-00001-of-00004.safetensors",
    "model.layers.0.mlp.act_fn.eps": "model-00001-of-00004.safetensors",
    "model.layers.0.mlp.down_proj.weight": "model-00001-of-00004.safetensors",
    "model.layers.0.mlp.up_proj.weight": "model-00001-of-00004.safetensors",
    "model.layers.0.self_attn.k_norm.weight": "model-00001-of-00004.safetensors",
    "model.layers.0.self_attn.k_proj.weight": "model-00001-of-00004.safetensors",
    "model.layers.0.self_attn.o_proj.weight": "model-00001-of-00004.safetensors",
    "model.layers.0.self_attn.q_norm.weight": "model-00001-of-00004.safetensors",
    "model.layers.0.self_attn.q_proj.weight": "model-00001-of-00004.safetensors",
    "model.layers.0.self_attn.v_proj.weight": "model-00001-of-00004.safetensors",
```

Looks exactly like the QK norm in Qwen3, a 128 flat tensor for both. But it somehow diverges from the Transformers implementation. Here's a dump from Transformers:

```console
ggml_debug: model.layers.0.self_attn.k_norm_in = (f32)  ... = {torch.Size([1, 8, 6, 128])}
                                     [
                                      [
                                       [      0.0260,       0.0009,      -0.0400, ...,       0.0216,       0.0396,      -0.0355]
                                       [     -0.0006,       0.0012,      -0.0057, ...,      -0.0108,      -0.0005,      -0.0019]
                                       [      0.0505,      -0.0432,       0.0101, ...,       0.0056,      -0.0203,       0.0132]
                                       [     -0.0055,      -0.0051,      -0.0081, ...,      -0.3099,      -0.0374,      -0.0569]
                                       [     -0.0056,      -0.0075,      -0.0123, ...,       0.1299,      -0.1645,      -0.0148]
                                       [      0.0052,      -0.0125,       0.0299, ...,       0.0775,      -0.1068,      -0.0833]
                                      ],
                                     ]
                                     sum = -4.458256

ggml_debug: model.layers.0.self_attn.k_norm_out = (f32)  ... = {torch.Size([1, 8, 6, 128])}
                                     [
                                      [
                                       [      1.6487,       0.0560,      -1.1489, ...,       0.8809,       1.1181,      -1.0577]
                                       [     -0.0343,       0.0733,      -0.1573, ...,      -0.5669,      -0.0163,      -0.0744]
                                       [      3.6375,      -3.0737,       0.3281, ...,       0.3351,      -0.8343,       0.5715]
                                       [     -0.2226,      -0.2069,      -0.1493, ...,     -10.3275,      -0.8633,      -1.3851]
                                       [     -0.1887,      -0.2521,      -0.1895, ...,       3.9971,      -3.5061,      -0.3333]
                                       [      0.3068,      -0.7309,       0.7994, ...,       3.6865,      -3.5230,      -2.8940]
                                      ],
                                     ]
                                     sum = 518.159729
```

and from llama-eval-callback:

```console
ggml_debug:                   Kcur-0 = (f32)    MUL_MAT(blk.0.attn_k.weight{4096, 1024, 1, 1}, attn_norm-0{4096, 6, 1, 1}}) = {1024, 6, 1, 1}
                                     [
                                      [
                                       [      0.0260,       0.0299,       0.0009, ...,       0.0173,       0.0058,      -0.0157],
                                       [      0.0602,       0.0272,      -0.0239, ...,       0.0578,       0.0108,      -0.0360],
                                       [      0.1024,       0.0396,       0.0892, ...,       0.0203,       0.0063,      -0.0092],
                                       [      0.0371,       0.0511,      -0.0486, ...,       0.0725,       0.0086,      -0.0277],
                                       [      0.0441,       0.0391,      -0.0433, ...,       0.0447,      -0.0085,      -0.0458],
                                       [      0.0566,       0.0338,      -0.0308, ...,       0.0555,       0.0202,      -0.0168],
                                      ],
                                     ]
                                     sum = -4.452065
ggml_debug:            Kcur_normed-0 = (f32)        MUL(CUDA0#norm-0#0{128, 8, 6, 1}, CUDA0#blk.0.attn_k_norm.weight#0{128, 1, 1, 1}}) = {128, 8, 6, 1}
                                     [
                                      [
                                       [      1.6491,       1.8751,       0.0248, ...,       3.8993,       0.2504,      -0.7591],
                                       [     -0.0332,      -0.1835,       0.0339, ...,       0.7072,       0.2774,       0.4394],
                                       [      3.6366,       2.1278,      -1.4065, ...,       0.4759,      -0.0021,      -0.0724],
                                       ..., 
                                       [      0.3081,      -0.1988,      -0.3355, ...,      -1.0101,      -0.1090,      -2.8183],
                                       [      2.0318,       0.6510,      -0.9407, ...,      -3.0910,      -0.1242,       1.1163],
                                       [     -1.1660,      -1.4152,       0.5272, ...,       1.8761,       0.4326,      -1.2444],
                                      ],
                                      [
                                       [      1.5329,       0.6841,      -0.2749, ...,       2.7403,      -0.5124,      -1.3517],
                                       [     -2.0542,      -1.4692,      -0.0407, ...,       1.1136,       0.5337,       0.1236],
                                       [      3.9727,       6.0690,      -2.1473, ...,      -2.2552,      -1.5008,      -0.9382],
                                       ..., 
                                       [     -0.2568,      -0.4579,      -0.3575, ...,       1.8921,       1.0777,      -3.7264],
                                       [      0.5295,       0.1762,      -0.1126, ...,      -3.3283,       0.6000,       1.7422],
                                       [     -5.0935,      -4.9426,       1.0877, ...,       2.7902,       0.3603,      -1.2684],
                                      ],
                                      [
                                       [      4.1699,       1.5921,       1.6440, ...,       1.3991,      -0.0073,      -1.1039],
                                       [     -0.0246,      -0.0408,      -0.0693, ...,      -0.2578,      -0.3472,       0.1711],
                                       [      0.2052,       0.0656,      -0.4119, ...,       0.1593,       0.0395,      -0.5277],
                                       ..., 
                                       [      1.1948,       0.1338,      -0.5707, ...,       5.8187,       0.8847,      -2.7834],
                                       [      0.4890,      -0.2239,      -0.0053, ...,      -2.8752,      -0.2403,       1.6677],
                                       [     -0.2667,      -0.4034,       0.6344, ...,       2.1922,       0.4676,      -0.7278],
                                      ],
                                      [
                                       [      0.9241,       1.2575,      -0.5471, ...,       3.1444,       0.3413,      -0.1515],
                                       [     -2.0858,      -0.2047,      -0.6143, ...,       0.7987,       1.1992,       0.1377],
                                       [      4.4787,       6.5776,      -1.9080, ...,      -1.2320,      -0.4378,       0.1753],
                                       ..., 
                                       [     -0.6235,      -0.1112,       0.2428, ...,       0.5462,       0.9830,      -2.6701],
                                       [      0.4527,       0.4643,      -0.3325, ...,      -2.3723,       0.1144,       1.5826],
                                       [     -5.4386,      -4.6627,       1.1660, ...,       3.1313,       0.2574,      -0.8740],
                                      ],
                                      [
                                       [      1.0395,       0.9112,      -0.4622, ...,       1.5698,      -0.1920,      -2.1213],
                                       [     -4.7884,      -2.8707,       0.0744, ...,       1.5661,       0.1393,      -0.9221],
                                       [      3.9333,       5.8780,      -1.7432, ...,       1.1649,       0.4673,      -0.8790],
                                       ..., 
                                       [     -0.8365,      -0.5343,       0.1528, ...,       1.2459,       0.5092,      -2.3725],
                                       [      0.1358,       0.4386,      -0.0570, ...,      -3.0326,       0.8687,       1.6394],
                                       [     -4.0204,      -3.5265,       0.9346, ...,       2.2506,      -0.2951,      -1.6851],
                                      ],
                                      [
                                       [      1.3533,       0.7988,      -0.3331, ...,       1.6144,       0.0773,      -1.0566],
                                       [      0.0273,      -0.4740,       0.0589, ...,      -0.0224,       0.2123,      -0.0747],
                                       [      4.2151,       5.8705,      -1.8385, ...,      -1.2067,       0.8555,       0.5714],
                                       ..., 
                                       [     -0.4960,       0.1881,       0.2486, ...,      -5.0892,      -0.0901,      -2.8932],
                                       [     -0.2326,       0.3233,       0.0688, ...,      -2.0252,      -0.3004,       1.8363],
                                       [     -5.1900,      -3.7914,       1.1631, ...,       2.5679,       0.6468,      -0.5669],
                                      ],
                                     ]
                                     sum = -38.386154
```

ggerganov commented:
Hm, not sure. Only guess is probably the Q and K tensors are permuted and need to be unpermuted during conversion?

pwilkin commented:
@ggerganov yeah, looks like it, now I'm trying to figure how exactly.

pwilkin commented:
Looks like some interleave is going on, because the first three elements of the first vector in GGML are: `0.0260,       0.0299,       0.0009`, which look like the first two elements of the first vector in Transformers, but with a value inserted between them. Not exactly sure what the operation is or why a totally standard op is suddenly misbehaving.

ggerganov commented:
Some models have permuted tensors which we reverse during the conversion. Take a look at this code for example:

https://github.com/ggml-org/llama.cpp/blob/3976dfbe00f02a62c0deca32c46138e4f0ca81d8/convert_hf_to_gguf.py#L1627-L1668

Maybe something similar has to be applied here too.

pwilkin commented:
My only worry is that there's nothing either in the Transformers implementation or the ChatLLM implementation (based on GGML after all) which suggests that there's any special treatment to be done. It seems like a normal extension of Llama3 with RMS norm. Even the tensor sizes look pretty standard. The conversion code is basically "extend Llama and add the extra tensors for the xIELU activation".

ngxson commented on src/llama-model.cpp:
```diff
@@ -18620,6 +18678,262 @@ struct llm_build_smallthinker : public llm_graph_context{
+
+                // xIELU activation
+                // Get the xIELU parameters from the model layers
+                ggml_tensor * alpha_n = model.layers[il].ffn_act_alpha_n;
+                ggml_tensor * alpha_p = model.layers[il].ffn_act_alpha_p;
+                ggml_tensor * beta = model.layers[il].ffn_act_beta;
+                ggml_tensor * eps = model.layers[il].ffn_act_eps;
+
+                float alpha_n_val = get_scalar_f32_val(alpha_n);
+                float alpha_p_val = get_scalar_f32_val(alpha_p);
+                float beta_val = get_scalar_f32_val(beta);
+                float eps_val = get_scalar_f32_val(eps);
```
This will end up being a very big performance penalty as there will be 4 extra round-trip CPU <--> backend each time a cgraph is constructed

A much simpler way is to read this as scalar value and write it as GGUF metadata upon conversion. This should also allow implementing xielu with existing GGML ops (no need to create new op)

pwilkin replied:
Yeah, was going to optimize the xIELU act as soon as I got a correct implementation working. You mean store them as vectors and load them as something like `n_head_arr`?

pwilkin commented:
Actually, the Q projection already diverges and I have no idea why... the attention norms are the same, but then I get this weird interleaving:

`[      **0.0287**,       0.0177,      **-0.0021**,      -0.0019,      **-0.0123**,      -0.0065,       0.0100...` for first tensor of Q (attn_q * attn_norm) in llama.cpp

`[      0.0287,      -0.0021,      -0.0123, ...` for first tensor of Q in Transformers

What could cause this sort of interleaving?

pwilkin commented:
Update: norm application is good after doing reshape pre-norm, now of course RoPE is wrong...

ggerganov commented:
> Update: norm application is good after doing reshape pre-norm, now of course RoPE is wrong...

Try to change the RoPE type to NEoX

pwilkin committed:
No permute during convert (fixes qk tensors), proper norm application.

pwilkin committed:
RoPE = NeoX

pwilkin committed:
Coherence!

pwilkin committed:
Merge branch 'ggml-org:master' into apertus-implementation

pwilkin committed:
Migrate xielu params from tensors to hyperparameters

pwilkin commented:
Aight, migrated the xIELU coefficients over to hyperparams, no new issues happily. Is there any way to optimally call xIELU using existing ops or do I just brute-force lift the operations to tensor level?

pwilkin commented:
For reference, I've tried:
* manually creating big filled tensors for constants
* implementing `expm1f` via custom_op_1 to see if it's the problem

theo77186 commented:
I tried different things: some algebraic manipulations on both the positive and negative branches of the function, and even using log1p for the softmax function. Nothing really changed. Also, what's the prompt used in `llama-eval-callback`?

pwilkin commented:
> In the xIELU paper they mention using expm1 for numerical stability.

Yeah, I explicitly implemented `expm1f` via `custom_op` for testing that hypothesis. But nothing really changed.

Also, I don't understand why `llama-eval-callback` finishes correctly and dumps the predictions while `llama-cli` fails. As I wrote above, the results from `llama-eval-callback` for the tensor vs scalar cases are extremely similar.

pwilkin commented:
Starting to think it might be some out-of-memory issue. Reduced layers offloaded to GPU, now I got:

```console
llama-server: /devel/tools/llama.cpp/src/llama-sampling.cpp:662: void llama_sampler_dist_apply(llama_sampler*, llama_token_data_array*): Assertion `found' failed.
Aborted (core dumped)
```

There's a funny comment there:
```c++
    // fallback to the last token (don't think this can happen)
```

theo77186 commented:
extra data point (with bf16 weights): `sum = -79.575645` when totally offloaded on GPU (CUDA, I believe this layer is on my 4060Ti)

theo77186 commented:
no crash, but when calculating perplexity, even with a context size of 128, I get NaN perplexity values

pwilkin committed:
Simple CUDA kernel

pwilkin commented:
Aight, reverted to the OP_XIELU implementation, added a simple CUDA kernel, performance seems passable:

ilintar@LinuksowaJaskinia:/mnt/win/k/models/ilintar/swiss-ai-Apertus-8B-Instruct-2509$ llama-bench -ngl 999 -fa 1 -m swiss-ai-Apertus-8B-Instruct-2509-q8_0.gguf 
load_backend: loaded BLAS backend from /devel/tools/llama.cpp/build/bin/libggml-blas.so
ggml_cuda_init: GGML_CUDA_FORCE_MMQ:    no
ggml_cuda_init: GGML_CUDA_FORCE_CUBLAS: no
ggml_cuda_init: found 1 CUDA devices:
  Device 0: NVIDIA GeForce RTX 3080, compute capability 8.6, VMM: yes
load_backend: loaded CUDA backend from /devel/tools/llama.cpp/build/bin/libggml-cuda.so
load_backend: loaded CPU backend from /devel/tools/llama.cpp/build/bin/libggml-cpu-haswell.so
| model                          |       size |     params | backend    | threads | fa |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ------: | -: | --------------: | -------------------: |
| apertus 8B Q8_0                |   7.97 GiB |     8.05 B | BLAS,CUDA  |       8 |  1 |           pp512 |       765.22 ± 19.15 |
| apertus 8B Q8_0                |   7.97 GiB |     8.05 B | BLAS,CUDA  |       8 |  1 |           tg128 |         46.30 ± 4.74 |

On to the chat template :)

CISC commented:
> Aight, reverted to the OP_XIELU implementation, added a simple CUDA kernel, performance seems passable:

Please revert the changes you did to the unary ops, this is wrong, what you are actually implementing is `XIELGLU`, which you should add to `GGML_GLU_OP`s, ie you just need to add a unary `op_xielu` (does not really need to be exposed as a GGML OP), and `ggml_compute_forward_xielglu_*`/`ggml_cuda_op_xielglu`.

theo77186 commented:
sorry if I misunderstood but the OP_XIELU as I read in the PR is not a gated linear unit, in fact there is no gated linear unit in Apertus at all. The various parameters are learnable scalars that makes the function vary non-linearly.

Performance is alright, though it seems RTX 30-series kinda struggle compared to 40-series.

CISC commented:
> sorry if I misunderstood but the OP_XIELU as I read in the PR is not a gated linear unit, in fact there is no gated linear unit in Apertus at all. The various parameters are learnable scalars that makes the function vary non-linearly.

You are right, I got confused by the `gate` references in the CUDA code, nevertheless the changes are not correct (also using `ggml_cuda_op_unary_gated` incorrectly, which was what confused me).

pwilkin commented:
@CISC @theo77186 Okay, wait, I feel like I need some explanation here (I am admittedly not very strong in the detailed algebraic side of ML)

Isn't xIELU basically a gated activation function? It's not linear, but it's gated (or more precisely, self-gated). So why is using `unary_gated_cuda wrong` here?

Also, why was the change to `unary-ops.cpp` / `unary.cu` wrong? It felt like a reasonable generalization of the current approach to lambdas with fixed captured parameters...

pwilkin commented:
> If you look more closely you will see that what it does (when given just `src[0]`) is split the input in two and use half of it as `ffn_gate`, ie. multiply it with the other half after applying the op on it.

Argh. I wonder how come it still worked though (as in, produced coherent results). 

> My issue was mainly with CUDA where you merged the inlines.

Yeah, that's what happens when you are trying to hurry and trust the LLM with "simple refactorings" :/

pwilkin committed:
Revert stupid LLM refactorings

pwilkin committed:
Chat template support

pwilkin committed:
configchecker / flake8 errors

mukel commented:
The model doesn't support thinking nor tools, despite being referenced in the chat template.

pwilkin commented:
> The model doesn't support thinking nor tools, despite being referenced in the chat template.

Thanks, that explains a few things 😄

CISC commented:
> Should be OK now

Nope, move the inlines back to the top, then you'll also notice that you've duplicated them for GLU.

pwilkin committed:
Reorder unary.cu

pwilkin committed:
I do conclude that LLMs are, in fact, stupid.

pwilkin commented:
> @pwilkin yes just want to confirm from SwissAI (I lead the reasoning and tooling) these are still training, the chat template was made to be compatible with a few things we're still working on, also feel free to message if you have any questions, I've forwarded about xIELU but haven't gotten a response yet

It's OK, the reasoning and tool calling support will be there for when they're ready.

mukel commented:
@YannickPezeu see https://github.com/ggml-org/llama.cpp/pull/15852#issuecomment-3271801568 , e.g. pass `--no-lazy` for the conversion.

pwilkin committed:
Merge branch 'master' into apertus-implementation

pwilkin committed:
Fix after merge

pwilkin committed:
Final newline

ggerganov commented on ggml/include/ggml.h:
```diff
@@ -554,6 +554,7 @@ extern "C" {
         GGML_OP_OPT_STEP_SGD,
 
         GGML_OP_GLU,
+        GGML_OP_XIELU,
```
Any reason not to put this along the rest of the unary ops like RELU, GELU, etc.?

pwilkin replied:
Not really :>

pwilkin replied:
@ggerganov Done.

pwilkin committed:
Make xIELU an UNARY_OP

pwilkin committed:
Final newline

ddh0 commented:
> I'm experiencing gibberish output (tested with q5) after commit https://github.com/ggml-org/llama.cpp/commit/2f68c03c55dabe5e7ab15c81c77ac9aa75753df7
Going back to that commit it seems fine (tested q5)
Can someone double check?

That commit you linked is just to add a single newline, so that's probably not the source of the issue.

pwilkin commented:
Aight I'll check, might have messed up the parameter indexing somewhere.

pwilkin committed:
Correctly account for parameter shift

pwilkin committed:
Argh.

pwilkin commented:
Ye, forgot to shift the params in the CPU implementation, should be fine now.

csabakecskemeti commented:
@pwilkin  I can confirm it's working now. I still experiencing  issues with the lower quants Q2 and Q3, above the output is reasonable

ggerganov commented on ggml/src/ggml-cpu/unary-ops.cpp:
```diff
@@ -64,8 +73,8 @@ static inline float op_log(float x) {
-template <float (*op)(float), typename src0_t, typename dst_t>
-static inline void vec_unary_op(int64_t n, dst_t * y, const src0_t * x) {
+template <typename Op, typename src0_t, typename dst_t>
+static inline void vec_unary_op(const Op& op, int64_t n, dst_t * y, const src0_t * x) {
```
```suggestion
static inline void vec_unary_op(const Op & op, int64_t n, dst_t * y, const src0_t * x) {
```

ggerganov commented on ggml/src/ggml.c:
```diff
@@ -2646,6 +2646,33 @@ struct ggml_tensor * ggml_silu(
     return ggml_unary(ctx, a, GGML_UNARY_OP_SILU);
 }
 
+// ggml_xielu
+static float softplus(float input) {
+    if (input > 20.0f) return input;
+    return logf(1 + expf(input));
+}
+
+struct ggml_tensor * ggml_xielu(
+        struct ggml_context * ctx,
+        struct ggml_tensor  * a,
+        float alpha_n,
+        float alpha_p,
+        float beta,
+        float eps) {
+    struct ggml_tensor * result = ggml_dup_tensor(ctx, a);
+
+    ggml_set_op_params_i32(result, 0, (int32_t) GGML_UNARY_OP_XIELU);
+    ggml_set_op_params_f32(result, 1, beta + softplus(alpha_n));
+    ggml_set_op_params_f32(result, 2, softplus(alpha_p));
+    ggml_set_op_params_f32(result, 3, beta);
+    ggml_set_op_params_f32(result, 4, eps);
+
+    result->op     = GGML_OP_UNARY;
+    result->src[0] = a;
+
+    return result;
+}
+
```
Move below after the `ggml_silu_inplace`

ggerganov commented on src/llama-model.cpp:
```diff
@@ -18792,6 +18852,154 @@ struct llm_build_smallthinker : public llm_graph_context{
     }
 };
 
+// TODO: maybe put this as a general helper in ggml.c?
+static float get_scalar_f32_val(const ggml_tensor *t) {
+    float onef;
+    if (t->buffer) {
+        ggml_backend_tensor_get(t, &onef, 0, sizeof(float));
+    } else {
+        GGML_ASSERT(t->data);
+        onef = *((float *) t->data);
+    }
+    return onef;
+}
+
```
Seems unused?

ggerganov reviewed (approved):
The model and CPU code is OK. The CUDA changes might need a second look (cc @JohannesGaessler / @am17an)

pwilkin committed:
Update ggml/src/ggml-cpu/unary-ops.cpp

Co-authored-by: Georgi Gerganov <ggerganov@gmail.com>

am17an commented on ggml/src/ggml-cuda/unary.cu:
```diff
@@ -276,29 +207,31 @@ void ggml_cuda_op_unary_gated(ggml_backend_cuda_context & ctx, ggml_tensor * dst
-    ggml_cuda_op_unary_gated<op_silu>(ctx, dst);
-}
+    __device__ __forceinline__ float operator()(float x) const {
+        float gate_pos = (x > 0.0f);        // positive branch gate
```
Use `const` wherever possible

pwilkin replied:
Done.

am17an commented on ggml/src/ggml-cuda/unary.cu:
```diff
@@ -276,29 +207,31 @@ void ggml_cuda_op_unary_gated(ggml_backend_cuda_context & ctx, ggml_tensor * dst
-void ggml_cuda_op_reglu(ggml_backend_cuda_context & ctx, ggml_tensor * dst) {
-    ggml_cuda_op_unary_gated<op_relu>(ctx, dst);
-}
+// Functor for XIELU operation with parameters
```
This comment is not necessary

pwilkin replied:
Removed.

am17an commented on ggml/src/ggml-cuda/unary.cu:
```diff
@@ -331,6 +264,158 @@ static void swiglu_oai_cuda(const T * x, const T * g, T * dst, const int64_t k,
+    leaky_relu_kernel<<<num_blocks, CUDA_RELU_BLOCK_SIZE, 0, stream>>>(x, dst, k, negative_slope);
+}
+
+// All ggml_cuda_op_* functions moved to bottom
```
It would be preferable to have this refactor in a separate PR

pwilkin replied:
Yeah, but I felt there was no good way to do this reasonably without refactoring the op abstraction, since otherwise I'd have to rewrite all the boilerplate around the unary op calls. Might be actually a good idea to rewrite all the other unary-ops-with-parameters as parameter-capturing lambdas to make it cleaner, unless there are specific optimizations there (I'm an absolute noob when it comes to CUDA programming, so maybe I'm not seeing something).

am17an commented on ggml/src/ggml.c:
```diff
@@ -2646,6 +2646,33 @@ struct ggml_tensor * ggml_silu(
+static float softplus(float input) {
+    if (input > 20.0f) return input;
+    return logf(1 + expf(input));
+}
```
This is also given here, might be better to abstract this out to a common class

https://github.com/ggml-org/llama.cpp/blob/e00f3fd8fff2cf5a8c8c9f475034bd089c8bcce4/ggml/src/ggml-cpu/ops.cpp#L9408-9409

pwilkin committed:
Refactor: remove unused methods, inline and factorize softplus, add const modifiers

ngxson commented on ggml/src/ggml-cpu/unary-ops.cpp:
```diff
@@ -95,25 +104,25 @@ static void apply_unary_op(const ggml_compute_params * params, ggml_tensor * dst
 // TODO: Use the 'traits' lookup table (for type conversion fns), instead of a mass of 'if' conditions with long templates
-template <float (*op)(float)>
-static void unary_op(const ggml_compute_params * params, ggml_tensor * dst) {
+template <typename Op>
+static void unary_op(const Op& op, const ggml_compute_params * params, ggml_tensor * dst) {
```
Instead of passing the function pointer here (which comes with the risk of performance degrade), you can simply change the `op` signature from `float (*op)(float)` to `float (*op)(float, ggml_tensor * dst)`

Having both `template` and `decltype(op)` here is pretty much redundant as they will be all compiled into one single version of the function, which against the reason why they use them in the first place.

accemlcc commented:
Hi Folks! Just saying -> 

✅ **SUCCESSFUL TESTING REPORT**

**Environment:**
- Windows 11, RTX 4090 24GB, CUDA 12.4.99
- Visual Studio 2022, conda environment
- Apertus 8B GGUF model

**Results:**
✅ Compilation successful (CPU + GPU)
✅ xIELU activation working correctly (α_n, α_p, β, ε parameters loaded)
✅ GPU acceleration: 33.50 tokens/sec on RTX 4090
✅ Flash Attention enabled, all 33 layers offloaded
✅ Complex generation: 509 tokens philosophical discussion
✅ Special tokens working: <|user_start|>, <|assistant_start|>, etc.

**Performance:** Matches PyTorch original quality and speed.
**Verdict:** Ready for merge! 🚀 <- Claude is it, thanks anyway!

theo77186 commented:
I've done some unscientific perplexity tests, seems like it really hates k-quants for some reason. Try using iq-quants, but these require imatrix as llama.cpp will simply refuse to make them for iq2 and iq3_xxs, and iq3 is seriously degraded without imatrix.

pwilkin commented:
I suspect `expm1f` is the culprit which makes the lower precision quants fail here.

csabakecskemeti commented:
FYI imatrix generation fails with
`compute_imatrix: computing over 93 chunks, n_ctx=512, batch_size=2048, n_seq=4
nan detected in blk.3.attn_q.weight
compute_imatrix: computing over 93 chunks, n_ctx=512, batch_size=2048, n_seq=4
nan detected in blk.3.attn_q.weight
`
f16.gguf has generated with the latest commit from this branch (--no-lazy is used)

csabakecskemeti commented:
IQ2_S produces reasonable output.
imatrixes (generated from Q8) if you want to play with it

- https://huggingface.co/DevQuasar/swiss-ai.Apertus-8B-Instruct-2509-GGUF/tree/main/imatrix
- https://huggingface.co/DevQuasar/swiss-ai.Apertus-70B-Instruct-2509-GGUF/tree/main/imatrix

pwilkin committed:
Merge branch 'master' into apertus-implementation

pwilkin committed:
Revert CUDA changes, implement xIELU as a separate OP

pwilkin committed:
Pesky newline

am17an reviewed (commented):
I think adding a test case for the xielu op in test-backend-ops would be nice, but I'm okay either way. I will leave @JohannesGaessler to approve

pwilkin committed:
Add float2half / half2float for F16 inputs/outputs

JohannesGaessler reviewed (approved):
Sorry, I think I may have looked at the wrong version. The CUDA changes as they are right now are fine.

JohannesGaessler commented:
(To be clear, the CI still needs to be fixed.)

pwilkin committed:
CUDA variants, attempt 2

pwilkin committed:
Actually, attempt 3

pwilkin commented:
@JohannesGaessler aight, I'm flailing in the dark a bit here, but I think I fixed the CUDA code now.

pwilkin committed:
Update ggml/src/ggml-cuda/unary.cu

Co-authored-by: Johannes Gäßler <johannesg@5d6.de>

pwilkin commented:
> Yes, going forward we should try to separate the implementation of new models better in order to simplify the review process.

Ye, will try to stick to the "CPU first, everything else later as a PR" model for the future, seems much easier to handle.

pwilkin committed:
Missing convert header

CISC commented on gguf-py/gguf/tensor_mapping.py:
```diff
@@ -1475,6 +1491,34 @@ class TensorNameMap:
                 "model.layers.{bid}.post_attention_layernorm",
             ),
         },
+        MODEL_ARCH.APERTUS: {
+            MODEL_TENSOR.ATTN_NORM: (
+                "model.layers.{bid}.attention_layernorm",
+            ),
+            MODEL_TENSOR.ATTN_Q_NORM: (
+                "model.layers.{bid}.attention.query_layernorm",
+                "model.layers.{bid}.self_attn.q_norm",
+            ),
+            MODEL_TENSOR.ATTN_K_NORM: (
+                "model.layers.{bid}.attention.key_layernorm",
+                "model.layers.{bid}.self_attn.k_norm",
+            ),
+            MODEL_TENSOR.FFN_NORM: (
+                "model.layers.{bid}.feedforward_layernorm",
+            ),
+            MODEL_TENSOR.FFN_ACT_ALPHA_N: (
+                "model.layers.{bid}.mlp.act_fn.alpha_n",
+            ),
+            MODEL_TENSOR.FFN_ACT_ALPHA_P: (
+                "model.layers.{bid}.mlp.act_fn.alpha_p",
+            ),
+            MODEL_TENSOR.FFN_ACT_BETA: (
+                "model.layers.{bid}.mlp.act_fn.beta",
+            ),
+            MODEL_TENSOR.FFN_ACT_EPS: (
+                "model.layers.{bid}.mlp.act_fn.eps",
+            ),
+        },
```
```suggestion
```
What is the reason for adding `NORM`s as arch-specific instead of regular mappings?

Make sure to add them after applying this.

slaren commented on ggml/src/ggml-impl.h:
```diff
@@ -102,6 +102,9 @@ static bool ggml_op_is_empty(enum ggml_op op) {
     }
 }
 
+static inline float softplus(float input) {
+    return (input > 20.0f) ? input : logf(1 + expf(input));
+}
```
This function doesn't belong here.

CISC replied:
@slaren gentle ping

pwilkin replied:
This is parameter prep, though, it's not really used in the backend calculations. I'll rename it.

pwilkin replied:
Done.

slaren commented on ggml/src/ggml.c:
```diff
@@ -2652,6 +2652,29 @@ struct ggml_tensor * ggml_silu_inplace(
+    struct ggml_tensor * result = ggml_dup_tensor(ctx, a);
+
+    ggml_set_op_params_i32(result, 0, (int32_t) GGML_UNARY_OP_XIELU);
+    ggml_set_op_params_f32(result, 1, beta + softplus(alpha_n));
+    ggml_set_op_params_f32(result, 2, softplus(alpha_p));
```
Why the changes to `alpha_n` and `alpha_p`?

pwilkin replied:
Added

slaren commented on ggml/src/ggml-cpu/unary-ops.cpp:
```diff
@@ -122,65 +131,80 @@ static void unary_op(const ggml_compute_params * params, ggml_tensor * dst) {
+
+void ggml_compute_forward_xielu(const ggml_compute_params * params, ggml_tensor * dst) {
+    // Get the XIELU parameters from the operation
+    float alpha_n = ggml_get_op_params_f32(dst, 1);
+    float alpha_p = ggml_get_op_params_f32(dst, 2);
```
Why do you read these values as `alpha_n` and `alpha_p` here when what is stored is something else?

slaren commented on ggml/include/ggml.h:
```diff
@@ -1148,6 +1149,16 @@ extern "C" {
             struct ggml_context * ctx,
             struct ggml_tensor  * a);
 
+    // xIELU activation function
+    // x = x * (alpha_n + alpha_p * sigmoid(beta * x)) + eps * (x > 0)
+    GGML_API struct ggml_tensor * ggml_xielu(
```
The comment does not seem to match the implementation.

pwilkin replied:
Added the constraining functions in the comments

slaren reviewed (commented):
Please document much better w

CISC commented on src/llama-model.cpp:
```diff
@@ -511,9 +511,13 @@ void llama_model::load_hparams(llama_model_loader & ml) {
-
     std::fill(hparams.swa_layers.begin(), hparams.swa_layers.end(), 0);
 
+    std::fill(hparams.xielu_alpha_n.begin(), hparams.xielu_alpha_n.end(), 0);
+    std::fill(hparams.xielu_alpha_p.begin(), hparams.xielu_alpha_p.end(), 0);
+    std::fill(hparams.xielu_beta.begin(), hparams.xielu_beta.end(), 0);
+    std::fill(hparams.xielu_eps.begin(), hparams.xielu_eps.end(), 0);
```
```suggestion
    std::fill(hparams.xielu_alpha_n.begin(), hparams.xielu_alpha_n.end(), 0.0f);
    std::fill(hparams.xielu_alpha_p.begin(), hparams.xielu_alpha_p.end(), 0.0f);
    std::fill(hparams.xielu_beta.begin(), hparams.xielu_beta.end(), 0.0f);
    std::fill(hparams.xielu_eps.begin(), hparams.xielu_eps.end(), 0.0f);
```

CISC commented on src/llama-model.cpp:
```diff
@@ -2007,6 +2011,19 @@ void llama_model::load_hparams(llama_model_loader & ml) {
             } break;
+        case LLM_ARCH_APERTUS:
+            {
+                ml.get_key(LLM_KV_ATTENTION_LAYERNORM_RMS_EPS, hparams.f_norm_rms_eps);
+                ml.get_key_or_arr(LLM_KV_XIELU_ALPHA_N,  hparams.xielu_alpha_n, hparams.n_layer);
+                ml.get_key_or_arr(LLM_KV_XIELU_ALPHA_P, hparams.xielu_alpha_p, hparams.n_layer);
+                ml.get_key_or_arr(LLM_KV_XIELU_BETA, hparams.xielu_beta, hparams.n_layer);
+                ml.get_key_or_arr(LLM_KV_XIELU_EPS, hparams.xielu_eps, hparams.n_layer);
```
```suggestion
                ml.get_key(LLM_KV_ATTENTION_LAYERNORM_RMS_EPS, hparams.f_norm_rms_eps);
                ml.get_key_or_arr(LLM_KV_XIELU_ALPHA_N,        hparams.xielu_alpha_n, hparams.n_layer);
                ml.get_key_or_arr(LLM_KV_XIELU_ALPHA_P,        hparams.xielu_alpha_p, hparams.n_layer);
                ml.get_key_or_arr(LLM_KV_XIELU_BETA,           hparams.xielu_beta,    hparams.n_layer);
                ml.get_key_or_arr(LLM_KV_XIELU_EPS,            hparams.xielu_eps,     hparams.n_layer);
```

CISC commented on src/llama-model.cpp:
```diff
@@ -5835,6 +5852,49 @@ bool llama_model::load_tensors(llama_model_loader & ml) {
+                        if (hparams.rope_scaling_type_train == LLAMA_ROPE_SCALING_TYPE_LONGROPE) {
+                            layer.rope_long  = create_tensor(tn(LLM_TENSOR_ROPE_FACTORS_LONG,  "weight", i), {n_rot/2}, TENSOR_NOT_REQUIRED | (i != 0 ? TENSOR_DUPLICATED : 0));
+                            layer.rope_short = create_tensor(tn(LLM_TENSOR_ROPE_FACTORS_SHORT, "weight", i), {n_rot/2}, TENSOR_NOT_REQUIRED | (i != 0 ? TENSOR_DUPLICATED : 0));
+                        }
+                        else {
```
```suggestion
                        } else {
```

CISC commented on src/llama-model.cpp:
```diff
@@ -5835,6 +5852,49 @@ bool llama_model::load_tensors(llama_model_loader & ml) {
+
+                    // output
+                    output_norm = create_tensor(tn(LLM_TENSOR_OUTPUT_NORM, "weight"), { n_embd }, 0);
+                    output      = create_tensor(tn(LLM_TENSOR_OUTPUT,      "weight"), {n_embd, n_vocab}, 0);
```
```suggestion
                    output      = create_tensor(tn(LLM_TENSOR_OUTPUT,      "weight"), { n_embd, n_vocab }, 0);
```

CISC commented on src/llama-model.cpp:
```diff
@@ -5835,6 +5852,49 @@ bool llama_model::load_tensors(llama_model_loader & ml) {
+                        layer.attn_norm = create_tensor(tn(LLM_TENSOR_ATTN_NORM, "weight", i), { n_embd }, 0);
+
+                        if (hparams.rope_scaling_type_train == LLAMA_ROPE_SCALING_TYPE_LONGROPE) {
+                            layer.rope_long  = create_tensor(tn(LLM_TENSOR_ROPE_FACTORS_LONG,  "weight", i), {n_rot/2}, TENSOR_NOT_REQUIRED | (i != 0 ? TENSOR_DUPLICATED : 0));
+                            layer.rope_short = create_tensor(tn(LLM_TENSOR_ROPE_FACTORS_SHORT, "weight", i), {n_rot/2}, TENSOR_NOT_REQUIRED | (i != 0 ? TENSOR_DUPLICATED : 0));
```
```suggestion
                            layer.rope_long  = create_tensor(tn(LLM_TENSOR_ROPE_FACTORS_LONG,  "weight", i), { n_rot/2 }, TENSOR_NOT_REQUIRED | (i != 0 ? TENSOR_DUPLICATED : 0));
                            layer.rope_short = create_tensor(tn(LLM_TENSOR_ROPE_FACTORS_SHORT, "weight", i), { n_rot/2 }, TENSOR_NOT_REQUIRED | (i != 0 ? TENSOR_DUPLICATED : 0));
```

CISC commented on src/llama-model.cpp:
```diff
@@ -5835,6 +5852,49 @@ bool llama_model::load_tensors(llama_model_loader & ml) {
+                            layer.rope_short = create_tensor(tn(LLM_TENSOR_ROPE_FACTORS_SHORT, "weight", i), {n_rot/2}, TENSOR_NOT_REQUIRED | (i != 0 ? TENSOR_DUPLICATED : 0));
+                        }
+                        else {
+                            layer.rope_freqs = create_tensor(tn(LLM_TENSOR_ROPE_FREQS, "weight", i), {n_rot/2}, TENSOR_NOT_REQUIRED | (i != 0 ? TENSOR_DUPLICATED : 0));
```
```suggestion
                            layer.rope_freqs = create_tensor(tn(LLM_TENSOR_ROPE_FREQS, "weight", i), { n_rot/2 }, TENSOR_NOT_REQUIRED | (i != 0 ? TENSOR_DUPLICATED : 0));
```

pwilkin committed:
Proper formula and reference for xIELU in the comments.

pwilkin committed:
Modify unary-ops.cpp to add the functor-based logic besides the template system to retain optimizations

pwilkin committed:
Apply suggestions from code review

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

pwilkin committed:
Add tensor mappings for Apertus to global list instead

pwilkin committed:
Fix lazy on scalars

pwilkin committed:
Merge remote-tracking branch 'pwilkin/master' into apertus-implementation

pwilkin commented:
Should be gtg.

CISC commented on convert_hf_to_gguf.py:
```diff
@@ -9107,7 +9144,7 @@ def meta_with_dtype_and_shape(cls, dtype: torch.dtype, shape: tuple[int, ...]) -
         dtype = cls._dtype_str_map[st_slice.get_dtype()]
         shape: tuple[int, ...] = tuple(st_slice.get_shape())
-        lazy = cls(meta=cls.meta_with_dtype_and_shape(dtype, shape), args=(st_slice,), func=lambda s: s[:])
+        lazy = cls(meta=cls.meta_with_dtype_and_shape(dtype, shape), args=(st_slice,), func=lambda s: s[...] if len(s.get_shape()) == 0 else s[:])
```
Why add this now, didn't you work around this already?

Probably belongs in a separate PR.

pwilkin replied:
It's a bugfix and I think it's too tiny for a separate PR anyway.

pwilkin committed:
Update ggml/src/ggml-cuda/unary.cu

Co-authored-by: Johannes Gäßler <johannesg@5d6.de>

pwilkin commented:
@CISC can we merge this?

CISC commented:
> @CISC can we merge this?

Most likely, let me ping on one outstanding change request first.

JohannesGaessler reviewed (approved):
CUDA changes LGTM.

pwilkin committed:
Add comment about the constraints on positive/negative alpha

pwilkin committed:
Change `softplus` to `ggml_softplus`

ChasingZenith commented:
I can't compile successfully. Possibly a bug is introduced.
https://github.com/ggml-org/llama.cpp/blob/128d522c04286e019666bd6ee4d18e3fbf8772e2/src/llama-model.cpp#L19266
```cpp
[ 15%] Built target llama-gguf-hash
/tmp/makepkg/llama.cpp-git/src/llama.cpp/src/llama-model.cpp: In constructor ‘llm_build_apertus::llm_build_apertus(const llama_model&, const llm_graph_params&)’:
/tmp/makepkg/llama.cpp-git/src/llama.cpp/src/llama-model.cpp:19266:43: error: ‘ggml_xielu’ was not declared in this scope; did you mean ‘ggml_silu’?
19266 |                 ggml_tensor * activated = ggml_xielu(ctx0, up, alpha_n_val, alpha_p_val, beta_val, eps_val);
      |                                           ^~~~~~~~~~
      |                                           ggml_silu
make[2]: *** [src/CMakeFiles/llama.dir/build.make:359: src/CMakeFiles/llama.dir/llama-model.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1053: src/CMakeFiles/llama.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

ThiloteE commented:
Llama.cpp commit 946f71ed9ade07e319859b5ce656144140e066fb
Just want to mention that it is recommended running the model on llama-server with `--jinja`. If not enabled, unwanted special tokens will show in responses.
